### PR TITLE
Fix copying byte arrays from storage to storage.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 1]
```diff
@@ -1633,8 +1633,9 @@ string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromTy
 
 				let oldLen := <byteArrayLength>(sload(slot))
 
+				let srcOffset := 0
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:1636]: This would not have been necessary, but I'm introducing it as preparation for memory objects.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 2]
```diff
@@ -1683,7 +1686,10 @@ string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromTy
 		templ("panic", panicFunction());
 		templ("byteArrayLength", extractByteArrayLengthFunction());
 		templ("dstDataLocation", arrayDataAreaFunction(_toType));
+		if (fromStorage)
+			templ("srcDataLocation", arrayDataAreaFunction(_fromType));
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:1690]: This is the actual fix.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 3]
```diff
@@ -1683,7 +1686,10 @@ string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromTy
...
 		templ("clearStorageRange", clearStorageRangeFunction(*_toType.baseType()));
+		templ("srcStride", to_string(fromStorage ? 1 : 0x20));
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp]: And this, too.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 4]
```diff
@@ -1652,22 +1653,24 @@ string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromTy
 				switch gt(newLen, 31)
 				case 1 {
 					let loopEnd := and(newLen, not(0x1f))
+					<?fromStorage> src := <srcDataLocation>(src) </fromStorage>
```
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.cpp:1656]: Wow... How this worked for bytes array longer than 31 elements. Probably no such test...

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 5]
```diff
@@ -1652,22 +1653,24 @@ string YulUtilFunctions::copyByteArrayToStorageFunction(ArrayType const& _fromTy
...
 					let dstPtr := dstDataArea
 					let i := 0
 					for { } lt(i, loopEnd) { i := add(i, 32) } {
```
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.cpp]: Maybe, to be consistent with the rest of this function?
```suggestion
					for { } lt(i, loopEnd) { i := add(i, 0x20) } {
```

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 3 above)
Comment (mijovic) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Not sure if we should call it `stride`, as stride is in bytes (bytes per one element), and for storage types we use `storageSize` when dealing with storage size. So maybe calling it `srcSize` here?

File: test/libsolidity/semanticTests/storage/storage_storgae_copy.sol [hunk 6]
```diff
@@ -0,0 +1,22 @@
+contract c {
```
Comment (mijovic) [inline test/libsolidity/semanticTests/storage/storage_storgae_copy.sol]: There is a typo in filename

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I'll call it `increment`

**Comment by chriseth** [comment]
Updated.

File: test/libsolidity/semanticTests/array/copying/bytes_storage_to_storage.sol [hunk 7]
```diff
@@ -0,0 +1,22 @@
+contract c {
+    bytes a;
+    bytes b;
+    function f(uint len) public returns (bytes memory) {
+        bytes memory x = new bytes(len);
+        for (uint i = 0; i < len; i++)
+            x[i] = byte(uint8(i));
+        a = x;
+        b = a;
```
Comment (mijovic) [inline test/libsolidity/semanticTests/array/copying/bytes_storage_to_storage.sol:11]: Maybe one more check:
```suggestion
        b = a;
        for (uint i = 0; i < len; i++)
            assert(a[i] == b[i]);
```

**Review by mijovic — COMMENTED** [review]
Also, would be great to have tests when we copy shorter byte array to longer and vice versa.
When we need to transit between different representations in both direction.

**Review by mijovic — APPROVED** [review]
Test you added has all the edge cases we need to test.

Giod to merge
